keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1452 @@
+However, as none of these are possible to guarantee, most service meshes (like
+Linkerd and Istio), use fragile and platform specific workarounds to have the basic functionality.
+For example, for termination, projects like
+[kubeexit](https://github.com/karlkfi/kubexit) or [custom solutions](https://suraj.io/post/how-to-gracefully-kill-kubernetes-jobs-with-a-sidecar/)
+are used. 
+
+Some service meshes depend on secret (like a certificate) downloaded by other
+Init container to establish secure communication between services. This makes
+the problem of ordering of sidecar and Init containers harder.
+
+### Problems: configuration / secrets
+
+Some pods use init containers to pull down configuration/secrets and update them
+before the main container gets it. Then use sidecars to continue to watch for
+changes and perform the updates and push to the main container. This requires
+two separate code paths today. Perhaps the same sidecar could be used for both
+cases.
+
+### Goals
+
+This proposal aims to:
+- make containers implementing the sidecar pattern first class citizens inside a
+  Pod
+- solve a Job completion issue when sidecars should run continuously
+- allow mixing initContainers and sidecars for a choreographed startup sequence
+- allow easy injection of sidecar containers in any Pod
+- (to be evaluated after alpha) allow to implement sidecar containers that will
+  guaranteed to be running longer than regular containers 
+
+### Non-Goals
+
+This proposal doesn't aim to:
+- support arbitrary dependencies graphs between containers
+- act as a security control to enforce that pod containers only run while the
+  sidecar is healthy. Restart of sidecar containers is a best effort
+- allow to enforce security boundaries to sidecar containers different that
+  other containers. For example, allow Istio to run privileged to configure ip
+  tables and disable other containers from doing so
+- (alpha) support containers termination ordering
+
+## Proposal
+
+The proposal is to introduce a `restartPolicy` field to Init containers and use
+it to indicate that an Init container is a sidecar container. Kubelet will start Init containers with
+`restartPolicy=Always` in the order with other Init containers, but instead of
+waiting for its completion, it will wait for the container startup completion.
+
+The condition for startup completion is open question. It will be either:
+- pod is ready (readiness probe succeeded) or
+- startup probe succeeded or if no startup probe defined, `postStart` handler
+completed. The decision on specific signal for container startup completion will
+be locked on before implementation will start. See the section
+["Pod startup completed condition"](#pod-startup-completed-condition)
+
+As part of the KEP, init containers and regular containers will be split into
+two different types. The field `restartPolicy` will only be introduced on init
+containers. The only supported value proposed in this KEP is `Always`. No other
```

Comment:
Would it be clearer to define this as having the values `""` and `Always`, rather than introducing a special null-able state?

Reply:
null value seems to be easier for rollout and rollback. nullable field is plan of record as of now